lxd/device/cdi/configure.go

```diff
@@ -0,0 +1,339 @@
+package cdi
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"golang.org/x/sys/unix"
+	"tags.cncf.io/container-device-interface/specs-go"
+
+	"github.com/canonical/lxd/lxd/instance"
+	"github.com/canonical/lxd/shared"
+	"github.com/canonical/lxd/shared/logger"
+)
+
+// specDevToNativeDev builds a list of unix-char devices to be created from a CDI spec.
+func specDevToNativeDev(configDevices *ConfigDevices, d specs.DeviceNode) error {
+	if d.Path == "" {
+		return fmt.Errorf("Device path is empty in the CDI device node: %v", d)
+	}
+
+	hostPath := d.HostPath
+	if hostPath == "" {
+		hostPath = d.Path // When the hostPath is empty, the path is the device path in the container.
+	}
+
+	if d.Major == 0 || d.Minor == 0 {
+		stat := unix.Stat_t{}
+		err := unix.Stat(hostPath, &stat)
+		if err != nil {
+			return err
+		}
+
+		d.Major = int64(unix.Major(uint64(stat.Rdev)))
+		d.Minor = int64(unix.Minor(uint64(stat.Rdev)))
+	}
+
+	configDevices.UnixCharDevs = append(configDevices.UnixCharDevs, map[string]string{"type": "unix-char", "source": hostPath, "path": d.Path, "major": fmt.Sprintf("%d", d.Major), "minor": fmt.Sprintf("%d", d.Minor)})
+	return nil
+}
+
+// specMountToNativeDev builds a list of disk mounts to be created from a CDI spec.
+func specMountToNativeDev(configDevices *ConfigDevices, cdiID ID, mounts []*specs.Mount) ([]SymlinkEntry, error) {
+	if len(mounts) == 0 {
+		return nil, fmt.Errorf("CDI mounts are empty")
+	}
+
+	indirectSymlinks := make([]SymlinkEntry, 0)
+	var chosenOpts []string
+
+	rootPath := ""
+	if shared.InSnap() {
+		rootPath = "/var/lib/snapd/hostfs"
+	}
+
+	for _, mount := range mounts {
+		if mount.HostPath == "" || mount.ContainerPath == "" {
+			return nil, fmt.Errorf("The hostPath or containerPath is empty in the CDI mount: %v", *mount)
+		}
```

Comment:
Note that strictly speaking it is possible to define a `tmpfs` which may have an empty `HostPath` and the `type` field set to `"tmpfs"`.  See https://github.com/opencontainers/runc/blob/346b818dad833a5ae8ab55d670b716dadd45950e/libcontainer/rootfs_linux.go#L521-L533

The source / host path is only required for a `bind` mount.